### ISSUE TITLE
Configurable attack frequency.

### ISSUE THREAD

**Issue by mibby** [kernitus/BukkitOldCombatMechanics#19]
Would it be possible to allow for a configurable increased attack frequency between hits for better hit detection / no damage ticks / no attack delay with a click-per-second limit to simulate faster combat similar to 1.8 in 1.10+?

Something like SmashHit (Async hit detection, but contains quite a few issues) https://aquifermc.org/resources/smashhit-async-hit-detection.15/ or MassiveCombat (has configurable damage ticks) https://www.spigotmc.org/threads/massivecombat-paid.151457/page-7#post-1924696

**Comment by gvlfm78** [comment]
I'm looking into this. What problems does SmashHit have?

**Comment by mibby** [comment]
https://github.com/PrimoMC/SmashHit/commit/fbbaa3293d5f5c7fac13a6d29aa5457c7de4f0b8#commitcomment-19449325

As well as issues commented on the forum discussion.

**Comment by gvlfm78** [comment]
Interesting. I'll see what can be done but I won't be able to handle this immediately. Also it cod initially have even more bugs than your alternative here but they'd probably be fixed in the testing phase, before being released.

**Comment by Drc-DEV** [comment]
You can already add this feature easily.

`
	public void onEnable() {
		if (ToggleHD) {
			for (Player p : Bukkit.getOnlinePlayers()) {
				setDamageTicks(p);
			}
		}
	}

        public static boolean ToggleHD = true;
        public static int HitD = 10;

	@EventHandler
	public void onPlayerRespawn(PlayerRespawnEvent e) {
		Player p = e.getPlayer();
		setDamageTicks(p);
	}

	@EventHandler
	public void onPlayerLogin(PlayerJoinEvent e) {
		Player p = e.getPlayer();
		setDamageTicks(p);
	}

	@EventHandler(ignoreCancelled = true)
	public void onPlayerChangeWorld(PlayerChangedWorldEvent e) {
		Player p = e.getPlayer();
		setDamageTicks(p);
	}

	public void setDamageTicks(Player p) {
		if (ToggleHD) {
			p.setMaximumNoDamageTicks(HitD);
		}
	}
`
Where ToggleHD is true/false based on configuration, and HitD is the ticks based on configuration.

**Comment by gvlfm78** [comment]
@Drc-DEV Would that do the trick? The example plugin shown had its own HashMap holding player attacks etc.

**Comment by Drc-DEV** [comment]
The HashMap holding player attacks is useful if dealing with "custom knockback", not hit delay. If you just want to add hit delay, that's it. @gvlfm78

**Comment by Rayzr522** [comment]
looks like this is actually more of a priority than I before thought, massively changing the dynamic of combat in 1.9+ vs 1.8.

there's a ton of information pertaining to this topic in this other duplicate ticket here: https://github.com/kernitus/BukkitOldCombatMechanics/issues/443#issuecomment-670344732

I'm bumping this feature request because I think this would be a big improvement to OCM, and seems to definitely be something needed by the PVP community

**Comment by kernitus** [comment]
Well, is the above code all that would be needed to implement this? Surely there's something more to it?

**Comment by Rayzr522** [comment]
i think at least for the basic functionality (configuring attack frequency), that's all we need as far as i'm aware. other related features like async hit detection would require packet-level work and much more in depth tweaking.

i think we could easily drop a quick 1.9.0 update with configurable attack frequency tho without too much work :)

**Comment by kernitus** [comment]
So... I have implemented code that sets the maximumNoDamageTicks for both players and mobs. However, I can't actually tell the difference. Using some debug statements I can see that the value has been changed, but I don't see any appreciable change to the actual noDamageTicks values for entities. @mibby @Drc-DEV @Rayzr522 You are welcome to try the [test build](https://ci.rayzr.dev/job/OldCombatMechanics/55/).
